pkg/v3/tickers/block.go

```diff
@@ -0,0 +1,74 @@
+package tickers
+
+import (
+	"context"
+	"log"
+	"sync"
+
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
+)
+
+type blockSubscriber interface {
+	// Subscribe provides an identifier integer, a new channel, and potentially an error
+	Subscribe() (int, chan ocr2keepers.BlockHistory, error)
+	// Unsubscribe requires an identifier integer and indicates the provided channel should be closed
+	Unsubscribe(int) error
+}
+
+// blockTicker is a struct that follows the same design paradigm as a time ticker but provides blocks
+// instead of time
+type blockTicker struct {
+	C          chan ocr2keepers.BlockHistory
+	chID       int
+	ch         chan ocr2keepers.BlockHistory
+	subscriber blockSubscriber
+	closer     sync.Once
+	ctx        context.Context
+	cancel     context.CancelFunc
+}
+
+func NewBlockTicker(subscriber blockSubscriber) (*blockTicker, error) {
+	chID, ch, err := subscriber.Subscribe()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	return &blockTicker{
+		chID:       chID,
+		ch:         ch,
+		C:          make(chan ocr2keepers.BlockHistory),
+		subscriber: subscriber,
+		closer:     sync.Once{},
+		ctx:        ctx,
+		cancel:     cancel,
+	}, nil
+}
+
+func (t *blockTicker) Start(ctx context.Context) (err error) {
+loop:
+	for {
+		select {
+		case blockHistory := <-t.ch:
+			select {
+			case t.C <- blockHistory:
```

Comment:
Is there a way we can store the last block history such that when the reading channel opens, we can read from the last instead of waiting for the next?

Reply:
So when t.C starts to reach from t.ch, you want t.C to immediately fetch whatever is in t.ch? So make t.ch a buffered channel of length 1?

Reply:
Kind of. I want the most recent item from `t.ch` to be pushed to `t.C` whenever `t.C` unblocks. So if the cadence of `t.C` reads is 4 seconds and the cadence of `t.ch` writes is 1.5 seconds, `t.ch` will write 2 times (3s) before `t.C` reads again. However, I don't want `t.C` to have to wait the extra 0.5 seconds for `t.ch` to write again.

If this is more complicated than it is worth to save 0.5 seconds, then maybe not bother. I'm just thinking of slower chains with 10+seconds on block cadence. It feels convoluted and maybe we can come back around to it.

Reply:
So like, am I right in thinking that t.ch should manage its own internal buffer, e.g.:

- If t.ch can send the current value to t.C, then send it
- If t.C is blocked, and t.ch wants to send a value, t.ch should overwrite an e.g. `nextValueBuffer` field, continuously for as long as t.C is blocked
- When t.C unblocks:
  - If t.ch has a value, then send it, and wipe `nextValueBuffer`
  - If t.ch doesn't have a value, then try to read from `nextValueBuffer`, if the value is set, then send it, otherwise do nothing until t.ch has something to send

Reply:
Or we could flip it so that the ticker always pulls into a temporary buffer and tries to send the value itself when t.ch has got nothing to send?

Reply:
Oh wait I'm not sure if we _can_ flip it, it might need to be buffered by t.ch?

Reply:
This looks right to me